### PR TITLE
Integrate feast direct storage 

### PR DESCRIPTION
**Pull request by tiopramayudi**
**What this PR does / why we need it**:

We already have feast direct storage implemented, but there is no way to orchestrate standard transformer using direct storage, hence this PR is introduced

**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

```release-note

```

**Checklist**

- [ ] Added unit test, integration, and/or e2e tests
- [ ] Tested locally
- [ ] Updated documentation
- [ ] Update Swagger spec if the PR introduce API changes
- [ ] Regenerated Golang and Python client if the PR introduce API changes

### CODE REVIEW

File: api/api/version_endpoints_api_test.go [hunk 1]
```diff
@@ -1950,10 +1950,286 @@ func TestCreateEndpoint(t *testing.T) {
...
-				DefaultFeastServingURL: "localhost:6566",
-				FeastServingURLs: config.FeastServingURLs{
-					{Host: "localhost:6566"},
-					{Host: "localhost:6567"},
+				FeastBigTableConfig: &config.FeastBigTableConfig{
+					ServingURL: "localhost:6567",
+				},
+				FeastRedisConfig: &config.FeastRedisConfig{
+					ServingURL: "localhost:6566",
+					RedisAddresses: []string{
```
Comment (pradithya) [inline api/api/version_endpoints_api_test.go:1958]: How will the `RedisAddresses` and other redis configuration will be configured?

Comment (pradithya) [reply api/api/version_endpoints_api_test.go:1958]: nvm found it

File: api/cluster/templater.go [hunk 2]
```diff
@@ -347,6 +321,50 @@ func (t *KFServingResourceTemplater) createTransformerSpec(modelService *models.
 	return transformerSpec
 }
 
+func (t *KFServingResourceTemplater) enrichStandardTransformerEnvVars(envVars models.EnvVars) models.EnvVars {
```
Comment (pradithya) [inline api/cluster/templater.go:324]: In this method there are 2 variables with almost same name (`standardTransformerConfig` and `t.standardTransformerConfig`) that has significantly different meaning. Can you change one of it.

File: api/cmd/api/ui_handler.go [hunk 3]
```diff
@@ -21,9 +21,9 @@ type uiEnvHandler struct {
 	DockerRegistries  string                `json:"REACT_APP_DOCKER_REGISTRIES,omitempty"`
 	MaxAllowedReplica int                   `json:"REACT_APP_MAX_ALLOWED_REPLICA,omitempty"`
 
-	DefaultFeastServingURL string                  `json:"REACT_APP_DEFAULT_FEAST_SERVING_URL,omitempty"`
-	FeastServingURLs       config.FeastServingURLs `json:"REACT_APP_FEAST_SERVING_URLS,omitempty"`
-	FeastCoreURL           string                  `json:"REACT_APP_FEAST_CORE_API,omitempty"`
+	DefaultFeastServingSource string                  `json:"REACT_APP_DEFAULT_FEAST_SOURCE,omitempty"`
```
Comment (pradithya) [inline api/cmd/api/ui_handler.go:24]: What would be the default value of it?

File: api/cmd/transformer/main.go [hunk 4]
```diff
@@ -137,47 +147,44 @@ func main() {
...
-func initFeastServingClients(feastOptions feast.Options, specs []*core.FeatureTableSpec, logger *zap.Logger) (feast.Clients, error) {
-	clients := feast.Clients{}
-
-	// Default Feast gRPC client. Used if user does not specify serving url
-	// on standard transformer config.
-	client, err := newFeastGrpcClient(feastOptions.DefaultServingURL)
-	if err != nil {
-		return nil, err
+func getServingType(userEnableDirectStorage *bool, currentServingType spec.ServingType) spec.ServingType {
+	servingType := currentServingType
```
Comment (pradithya) [inline api/cmd/transformer/main.go:151]: Can you simplify this part. It seems overly complicated where simple `if else` statement will do
```
if userEnableDirectStorage == nil {
  return currentServingType
}

if userEnableDirectStorage {
  return spec.ServingType_DIRECT_STORAGE
}

return spec.ServingType_FEAST_GRPC
```

File: api/cmd/transformer/main.go [hunk 5]
```diff
@@ -137,47 +147,44 @@ func main() {
...
-	clients[feast.URL(feast.DefaultClientURLKey)] = client
-	logger.Info("Default Feast gRPC client initialized", zap.String("url", feastOptions.DefaultServingURL))
+	return servingType
+}
 
-	for _, url := range feastOptions.ServingURLs {
-		client, err := newFeastGrpcClient(url)
-		if err != nil {
-			return nil, err
+func overwriteFeastOptionsConfig(appConfig AppConfig) feast.Options {
```
Comment (pradithya) [inline api/cmd/transformer/main.go:159]: Can you add comments what is the purpose of this function

File: api/config/config.go [hunk 6]
```diff
@@ -191,21 +192,24 @@ type MlpAPIConfig struct {
...
-	FeastCoreAuthAudience  string           `envconfig:"FEAST_CORE_AUTH_AUDIENCE" required:"true"`
-	EnableAuth             bool             `envconfig:"FEAST_AUTH_ENABLED" default:"false"`
-	Jaeger                 JaegerConfig
+	ImageName             string               `envconfig:"STANDARD_TRANSFORMER_IMAGE_NAME" required:"true"`
+	FeastServingURLs      FeastServingURLs     `envconfig:"FEAST_SERVING_URLS" required:"true"`
+	FeastCoreURL          string               `envconfig:"FEAST_CORE_URL" required:"true"`
+	FeastCoreAuthAudience string               `envconfig:"FEAST_CORE_AUTH_AUDIENCE" required:"true"`
+	EnableAuth            bool                 `envconfig:"FEAST_AUTH_ENABLED" default:"false"`
+	FeastRedisConfig      *FeastRedisConfig    `envconfig:"FEAST_REDIS_CONFIG"`
+	FeastBigTableConfig   *FeastBigTableConfig `envconfig:"FEAST_BIG_TABLE_CONFIG" required:"true"`
```
Comment (pradithya) [inline api/config/config.go:201]: maybe we shouldn't set `FeastBigTableConfig` as `required: "true"`

File: api/config/feast_storage_config.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+			return err
+		}
+		*d = Duration(timeDuration)
+		return nil
+	default:
+		return errors.New("invalid duration")
+	}
+}
+
+type FeastRedisConfig struct {
```
Comment (pradithya) [inline api/config/feast_storage_config.go:44]: Can you add description for all the fields

File: api/pkg/transformer/feast/feature_retriever.go [hunk 8]
```diff
@@ -86,10 +86,21 @@ func NewFeastRetriever(
 	}
 }
 
+type FeastStorageConfig map[spec.ServingSource]*spec.OnlineStorage
```
Comment (pradithya) [inline api/pkg/transformer/feast/feature_retriever.go:89]: Can we put it in separate file

File: api/pkg/transformer/feast/feature_retriever.go [hunk 9]
```diff
@@ -106,12 +117,39 @@ type Options struct {
 	FeastClientErrorPercentThreshold  int           `envconfig:"FEAST_HYSTRIX_ERROR_PERCENT_THRESHOLD" default:"25"`
 }
 
-func (o Options) IsServingURLSupported(url string) bool {
-	for _, supportedURL := range o.ServingURLs {
-		if supportedURL == url {
+func (o Options) IsFeastConfigUsingCorrectSource(featureTableCfg *spec.FeatureTable) bool {
```
Comment (pradithya) [inline api/pkg/transformer/feast/feature_retriever.go:120]: I think this should be part of `validator.go` and the method should accept the `Options` as parameter not as receiver.  i.e.
```
func IsFeastConfigUsingCorrectSource(option Options, featureTableCfg *spec.FeatureTable) bool {
 ...
}
```

Comment (pradithya) [reply api/pkg/transformer/feast/feature_retriever.go:120]: The name also need to be improved to reflect better about what the function does.

File: api/pkg/transformer/feast/feature_table.go [hunk 10]
```diff
@@ -0,0 +1,98 @@
+package feast
```
Comment (pradithya) [inline api/pkg/transformer/feast/feature_table.go:1]: Can you add documentation to all the functions introduced within `feature_table.go` .

File: api/pkg/transformer/feast/storage.go [hunk 11]
```diff
@@ -6,24 +6,23 @@ import (
...
 	"github.com/spaolacci/murmur3"
 	"google.golang.org/protobuf/types/known/durationpb"
 	"google.golang.org/protobuf/types/known/timestamppb"
 )
 
-func NewDirectStorageClient(storage *OnlineStorage, featureTables []*core.FeatureTableSpec) (StorageClient, error) {
+func NewDirectStorageClient(storage *spec.OnlineStorage, featureTablesMetadata []*spec.FeatureTableMetadata) (StorageClient, error) {
 	switch storage.Storage.(type) {
-	case *OnlineStorage_Redis:
+	case *spec.OnlineStorage_Redis:
```
Comment (pradithya) [inline api/pkg/transformer/feast/storage.go:25]: We have only 2 types of `ServingSource`, how do we differentiate between redis and redis cluster? @tiopramayudi @khorshuheng

File: api/service/version_endpoint_service.go [hunk 12]
```diff
@@ -112,8 +124,17 @@ func (k *endpointService) DeployEndpoint(environment *models.Environment, model
 			newEndpoint.Transformer.TransformerType = models.CustomTransformerType
 		}
 
+		if newEndpoint.Transformer.TransformerType == models.StandardTransformerType {
```
Comment (pradithya) [inline api/service/version_endpoint_service.go:127]: Can you add comment why `reconfigureStandardTransformer` is needed?

File: protos/merlin/transformer/spec/feast.proto [hunk 13]
```diff
@@ -4,12 +4,20 @@ package merlin.transformer;
 
 option go_package = "github.com/gojek/merlin/pkg/transformer/spec";
 
+import "google/protobuf/duration.proto";
```
Comment (pradithya) [inline protos/merlin/transformer/spec/feast.proto:7]: Can you add documentation to all proto fields

File: protos/merlin/transformer/spec/storage.proto [hunk 14]
```diff
@@ -2,15 +2,21 @@ syntax = "proto3";
 
 package merlin.transformer;
 
-option go_package = "github.com/gojek/merlin/pkg/transformer/feast";
+option go_package = "github.com/gojek/merlin/pkg/transformer/spec";
 
 import "google/protobuf/duration.proto";
 
+enum ServingType {
```
Comment (pradithya) [inline protos/merlin/transformer/spec/storage.proto:9]: Same as https://github.com/gojek/merlin/pull/195/files#r749074194

File: python/sdk/test/integration_test.py [hunk 15]
```diff
@@ -838,17 +838,125 @@ def test_standard_transformer_with_multiple_feast(
...
         }
     }
 
+
+    assert resp.json()["instances"] == exp_resp["instances"]
+    merlin.undeploy(v)
+
+@pytest.mark.feast
+@pytest.mark.integration
+def test_standard_transformer_with_multiple_feast_with_source(
```
Comment (pradithya) [inline python/sdk/test/integration_test.py:875]: Will it test the direct storage approach?

File: ui/src/pages/version/components/forms/components/feast_config/components/FeastServingUrlSelect.js [hunk 16]
```diff
@@ -15,10 +15,33 @@ const logoMaps = {
   bigtable: logoBigtable
 };
 
-export const FeastServingUrlSelect = ({ servingUrl, onChange, isInvalid }) => {
+export const FeastServingUrlSelect = ({
```
Comment (pradithya) [inline ui/src/pages/version/components/forms/components/feast_config/components/FeastServingUrlSelect.js:18]: Can you update the PR description with the screenshot of this change

File: api/cmd/transformer/main.go (hunk 4 above)
Comment (tiopramayudi) [reply api/cmd/transformer/main.go:151]: just realised it, will change

File: api/cmd/api/ui_handler.go (hunk 3 above)
Comment (tiopramayudi) [reply api/cmd/api/ui_handler.go:24]: BIGTABLE

File: api/pkg/transformer/feast/storage.go (hunk 11 above)
Comment (tiopramayudi) [reply api/pkg/transformer/feast/storage.go:25]: it depends on how orchestrator(merlin api) setup feast config. In standard transformer merlin will choose one of redis or redis cluster for REDIS source to pass to standard transformer.

Comment (tiopramayudi) [reply api/pkg/transformer/feast/storage.go:25]: in this PR I only consider redis cluster, but I'll update it to acommodate user to use single redis storage

File: python/sdk/test/integration_test.py (hunk 15 above)
Comment (tiopramayudi) [reply python/sdk/test/integration_test.py:875]: yes

File: ui/src/pages/version/components/forms/components/feast_config/components/FeastServingUrlSelect.js (hunk 16 above)
Comment (tiopramayudi) [reply ui/src/pages/version/components/forms/components/feast_config/components/FeastServingUrlSelect.js:18]: there is no change on UI look

**Comment by pradithya** [comment]
LGTM

**Review by pradithya — APPROVED** [review]

**Pull request merged**